p2p/dial.go

```diff
@@ -194,42 +145,45 @@ func (s *dialstate) newTasks(nRunning int, peers map[enode.ID]*Peer, now time.Ti
 			newtasks = append(newtasks, t)
 		}
 	}
+
+	// Compute number of dynamic dials needed.
+	needDynDials := s.maxDynDials
+	for _, p := range peers {
+		if p.rw.is(dynDialedConn) {
+			needDynDials--
+		}
+	}
+	for _, flag := range s.dialing {
+		if flag&dynDialedConn != 0 {
+			needDynDials--
+		}
+	}
+
 	// If we don't have any peers whatsoever, try to dial a random bootnode. This
 	// scenario is useful for the testnet (and private networks) where the discovery
 	// table might be full of mostly bad peers, making it hard to find good ones.
 	if len(peers) == 0 && len(s.bootnodes) > 0 && needDynDials > 0 && now.Sub(s.start) > fallbackInterval {
 		bootnode := s.bootnodes[0]
 		s.bootnodes = append(s.bootnodes[:0], s.bootnodes[1:]...)
 		s.bootnodes = append(s.bootnodes, bootnode)
-
 		if addDial(dynDialedConn, bootnode) {
 			needDynDials--
 		}
 	}
-	// Use random nodes from the table for half of the necessary
-	// dynamic dials.
-	randomCandidates := needDynDials / 2
-	if randomCandidates > 0 {
-		n := s.ntab.ReadRandomNodes(s.randomNodes)
-		for i := 0; i < randomCandidates && i < n; i++ {
-			if addDial(dynDialedConn, s.randomNodes[i]) {
```

Comment:
I don't get the rationale behind this change. Why an outbound peer cannot be select from the table?
Isn't that relying on the random walk only may not end up finding a good/reachable peer?

Reply:
I removed the call here because this mechanism (ReadRandomNodes) is very specific to the discovery package. Using just the iterator means we can use any iterator implementation here. 

The RandomNodes iterator in p2p/discover does return table nodes at the beginning of each lookup, so we're not really losing those local results.

Reply:
@fjl Can you point me to the code that RandomNodes iterator returns table nodes?
I am reviewing the latest release and RandomNodes at p2p/discover/v4_udp only returns the results of a lookup to a random 64-byte target string (refers to newRandomLookup).
I also thought the only thing this commit should change is the node iterator for the randomLookup, not the ReadRandomNodes.